net: conn_mgr: Add IPv4 and IPv6 tracking

glarsennordic opened the pull request.

glarsennordic committed:
net: conn_mgr: Track ready count ephemerally

Instead of incrementing and decrementing global counter,
just recompute the ready-count from scratch every time
conn_mgr_mon_handle_update is called.

This will simplify the introduction of additional ready count types.

This should have no externally observable impact on the behavior of
conn_mgr.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Add IPv4 and IPv6 tracking

conn_mgr now fires:

- NET_EVENT_L4_IPV4_CONNECTED
- NET_EVENT_L4_IPV4_DISCONNECTED
- NET_EVENT_L4_IPV6_CONNECTED
- NET_EVENT_L4_IPV6_DISCONNECTED

These events track whether there are any ready ifaces offering
specifically IPv4 or specifically IPv6 connectivity.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented:
I haven't added unit tests for the new feature, nor have I updated old unit tests. I am going to put that off until after I transition my work laptop to linux, since it will be much easier to do local testing in that environment

glarsennordic committed:
test: net: conn_mgr: Add IPv4/IPv6 notif test coverage

Modify the test cycle routines to verify recently added IPv4- and
IPv6-specific events.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Simplify blame handling

Don't track up/down blame separately.
Instead, track a single last-blame iface.

Don't track blame iface inside set_ready.
Instead, track directly inside handle_update.

These two changes will simplify the addition
of blame for IPv4- and IPv6-specific events.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented:
@simensrostad I cannot add you as reviewer, but please feel free to review

simensrostad reviewed (approved):
Looked over new events. looks good

rlubos commented on include/zephyr/net/conn_mgr_connectivity.h:
```diff
@@ -60,6 +65,30 @@ enum net_event_conn_cmd {
+/**
+ * @brief net_mgmt event raised when IPv4 connectivity becomes available
+ */
+#define NET_EVENT_CONN_IPV4_CONNECTED					\
```
Given those events are similar to `NET_EVENT_L4_CONNECTED` etc., shouldn't they be grouped together, i. .e be specified in https://github.com/zephyrproject-rtos/zephyr/blob/0b220dd6a6d1a4320e07071d1bdc69adcbe6f65b/include/zephyr/net/net_event.h#L281
? Not only we end up with those apparently related events scattered across the codebase, but also given that they use a different event base, it won't be possible to combine those events in a single event handler.

glarsennordic replied:
Yeah, it would definitely be nicer to have all of them in one event group, I agree

My reason for separating them was that, back when we were discussing how to implement `conn_mgr`, one of the issues that came up was that we felt that `NET_EVENT_L4_CONNECTED` didn't really fit well in the `NET_EVENT_L4` group. But when we tried to move it, we got pushback since it was a breaking change, and decided to just leave it there.

Personally I'm neutral to where it goes, so if you think we should just stick the new events into `NET_EVENT_L4`, I'm happy to do so. It would definitely be more convenient to have them all under one roof

jukkar replied:
I agree with @rlubos, better to have the L4 events in one place.

rlubos reviewed (commented):
Looks good overall, just one concern to discuss.

jukkar reviewed (commented):
LGTM, small nit about the commit message which has old information.

jukkar commented on include/zephyr/net/net_event.h:
```diff
@@ -289,6 +293,23 @@ enum net_event_l4_cmd {
 
+
+/** Event raised when IPv4 network connectivity becomes available. */
+#define NET_EVENT_L4_IPV4_CONNECTED			\
```
Could you also add these new events to `subsys/net/lib/shell/events.c` so that the event monitor can show user friendly information about them.

glarsennordic committed:
net: shell: Add IPv4 and IPv6 connectivity events

Add descriptions for recently introduced IPv4 and IPv6
connectivity events to the net event monitor.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -73,6 +73,17 @@ Afterwards, ifaces can become ready or unready without firing additional events,
 
```
Good catch

kartben commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -73,6 +73,17 @@ Afterwards, ifaces can become ready or unready without firing additional events,
+
+   Connection Manager also fires the following more specific ``CONNECTED`` / ``DISCONNECTED`` events:
+
+   - :c:macro:`NET_EVENT_L4_IPV4_CONNECTED`
+   - :c:macro:`NET_EVENT_L4_IPV4_DISCONNECTED`
+   - :c:macro:`NET_EVENT_L4_IPV6_CONNECTED`
+   - :c:macro:`NET_EVENT_L4_IPV6_DISCONNECTED`
```
```suggestion
   - :c:enumerator:`NET_EVENT_L4_IPV4_CONNECTED`
   - :c:enumerator:`NET_EVENT_L4_IPV4_DISCONNECTED`
   - :c:enumerator:`NET_EVENT_L4_IPV6_CONNECTED`
   - :c:enumerator:`NET_EVENT_L4_IPV6_DISCONNECTED`
```

glarsennordic replied:
Are you certain these should be changed to `:c:enumerator:`? They are defined as macros whose values derive from the actual enumerator values

kartben replied:
Of course! Read too fast and missed the difference between the "CMD" enumerators ans the macros

kartben commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -73,6 +73,17 @@ Afterwards, ifaces can become ready or unready without firing additional events,
+   - :c:macro:`NET_EVENT_L4_IPV6_CONNECTED`
+   - :c:macro:`NET_EVENT_L4_IPV6_DISCONNECTED`
+
+   These are similar to :c:macro:`NET_EVENT_L4_CONNECTED` and :c:macro:`NET_EVENT_L4_DISCONNECTED`, but specifically track whether IPv4- and IPv6-capable ifaces are ready.
```
```suggestion
   These are similar to :c:enumerator:`NET_EVENT_L4_CONNECTED` and :c:enumerator:`NET_EVENT_L4_DISCONNECTED`, but specifically track whether IPv4- and IPv6-capable ifaces are ready.
```

kartben commented on include/zephyr/net/net_event.h:
```diff
@@ -289,6 +293,23 @@ enum net_event_l4_cmd {
+#define NET_EVENT_L4_IPV4_CONNECTED			\
+	(_NET_EVENT_L4_BASE | NET_EVENT_L4_CMD_IPV4_CONNECTED)
+
+/** Event emitted when IPv4 network connectivity becomes lost. */
```
```suggestion
/** Event emitted when IPv4 network connectivity is lost. */
```

kartben commented on include/zephyr/net/net_event.h:
```diff
@@ -289,6 +293,23 @@ enum net_event_l4_cmd {
+#define NET_EVENT_L4_IPV6_CONNECTED			\
+	(_NET_EVENT_L4_BASE | NET_EVENT_L4_CMD_IPV6_CONNECTED)
+
+/** Event emitted when IPv6 network connectivity becomes available. */
```
```suggestion
/** Event emitted when IPv6 network connectivity is lost. */
```

glarsennordic replied:
Good catch

kartben reviewed (changes requested):
some minor doc issues and an actual comment typo

kartben reviewed (changes requested):
some minor doc issues and an actual comment typo

kartben reviewed (approved):
Doc ok, thx!
